packages/sdk-socket-server-next/src/redis.ts

```diff
@@ -0,0 +1,492 @@
+  );
+
+  // Remove these individual event logs to reduce noise
+  // These events still happen but we don't log each occurrence
+  newRedisClient.on('connect', () => {
+    // Silent connection
+  });
+
+  newRedisClient.on('close', () => {
+    // Silent close
+  });
+
+  newRedisClient.on('reconnecting', () => {
+    // Silent reconnection
+  });
+
+  newRedisClient.on('end', () => {
+    // Silent end
+  });
+
+  return newRedisClient;
+};
+
+const redisFactory = {
+  create: async () => {
+    // Create a unique key for this client
+    const cacheKey = `redis-client-${Date.now()}-${Math.random()
+      .toString(36)
+      .substring(2, 15)}`;
+
+    // Only log once per 50 clients to reduce noise (increased from 10)
+    const shouldLog =
+      redisClientCache.size % 50 === 0 || redisClientCache.size === 0;
+    if (shouldLog) {
+      logger.info(
+        `Redis pool: Creating client (cache size: ${redisClientCache.size})`,
+      );
+    }
+
+    const client = buildRedisClient(false);
+    if (!client) {
+      // Handle case where client couldn't be built (e.g., no nodes)
+      logger.error('Failed to create Redis client in factory');
+      // Depending on desired behavior, you might throw an error or return a mock/null client
+      // For now, let's throw to make the issue explicit
+      throw new Error('Failed to build Redis client in pool factory');
+    }
+    redisClientCache.set(cacheKey, client);
+
+    // Add client-specific reference to allow cleanup
+    (client as any).__cacheKey = cacheKey;
+
+    return client; // Resolve with the client directly
+  },
+  destroy: (client: Cluster | Redis) => {
+    // Get cache key from client if available
+    const cacheKey = (client as any).__cacheKey;
+    if (cacheKey) {
+      redisClientCache.delete(cacheKey);
+    }
+
+    // Only log once per 50 clients to reduce noise (increased from 10)
+    const shouldLog =
+      redisClientCache.size % 50 === 0 || redisClientCache.size === 0;
+    if (shouldLog) {
+      logger.info(
+        `Redis pool: Destroying client (cache size: ${redisClientCache.size})`,
+      );
+    }
+
+    return Promise.resolve(client.disconnect());
+  },
+};
+
+let redisClient: Cluster | Redis | undefined;
+
+export const getGlobalRedisClient = () => {
+  if (!redisClient) {
+    redisClient = buildRedisClient();
+  }
+
+  // Ensure redisClient is defined before returning
+  if (!redisClient) {
+    throw new Error('Global Redis client could not be initialized.');
+  }
+  return redisClient;
+};
+
+export const pubClientPool = genericPool.createPool(redisFactory, {
+  max: 35,
+  min: 15,
+  acquireTimeoutMillis: 15000,
+  idleTimeoutMillis: 300000,
+  evictionRunIntervalMillis: 180000,
+  numTestsPerEvictionRun: 2,
+  softIdleTimeoutMillis: 240000,
+});
+
+/**
+ * PooledClientWrapper - A Redis client wrapper that uses the connection pool internally
+ * ... (rest of the class definition as before) ...
+ */
+class PooledClientWrapper {
```

Comment:
While it does seem easier to work with, this seems odd as a pattern. If any of the protocol messages use `pubClient` more than once, then they may be using different redis clients between different calls (`get`, `set`, etc.), which seems inefficient at best or that it would cause concurrency issues at worse. But since pooled clients don't use auto pipelining, maybe the concurrency issues aren't an actual concern and just the use of multiple clients per call is more of the concern